modules/GenomicsEnglandPanelApp.py

```diff
@@ -110,53 +154,113 @@ def publication_from_symbol(symbol, response):
+                            "Phenotypes",
+                            when(col("Phenotypes") == "No OMIM phenotype",
+                                col("Panel Name"))
+                            .when(col("Phenotypes").isNull(),
+                                col("Panel Name"))
+                            .otherwise(col("Phenotypes"))
+                        )
+                        # cohortPhenotypes --> array of the original string separated by phenotypes 
+                        .withColumn(
+                            "cohortPhenotypes",                          
+                            split(col("Phenotypes"), ";")
+                        )
+                        # phenotype --> explosion of cohortPhenotypes
+                        .withColumn(
+                            "phenotype",
+                            explode(col("cohortPhenotypes"))
+                        )
+                        # omimCode --> OMIM code present in "phenotype"
+                        .withColumn("omimCode", regexp_extract(col("phenotype"), "(\d{6})", 1))
+                        # removal of the OMIM code in "phenotype"
+                        .withColumn("phenotype", regexp_replace(col("phenotype"), "(\d{6})", ""))
+                        # deleting special characters in "phenotype"
+                        .withColumn("phenotype", regexp_replace(col("phenotype"), "[^0-9a-zA-Z *]", ""))
+                        .withColumn("phenotype", trim(col("phenotype")))
+        )
 
-        #dataframe["Phenotype"] = dataframe["Phenotype"].str.replace("-", " ", regex=False) 
-        dataframe["Phenotype"] = dataframe["Phenotype"].str.replace("[^0-9a-zA-Z *]", "", regex=True)
-        dataframe["Phenotype"] = dataframe["Phenotype"].apply(lambda X: X.strip())
+        return dataframe
+    
+    def diseaseMappingStep(self):
+        '''
+        Runs the disease mapping step from phenotypes and OMIM codes to an EFO ID - 3 steps:
+            1. Querying OnToma with all distinct codes and phenotypes
+            2. Xref between the results of every phenotype/OMIM code pair for a better coverage
+            3. Mappings are built into the dataframe
 
-        dataframe.reset_index(inplace=True)
+        Args:
+            dataframe (pyspark.DataFrame): DataFrame with transformed PanelApp data
+        Returns:
+            dataframe (pyspark.DataFrame): DataFrame with the mapping results filtered by only matches
+        '''
 
-        return dataframe
+        omimCodesDistinct = list(self.dataframe.select("omimCode").distinct().toPandas()["omimCode"])
+        phenotypesDistinct = list(self.dataframe.select("phenotype").distinct().toPandas()["phenotype"])
+        omimCodes = list(self.dataframe.toPandas()["omimCode"])
+        phenotypes = list(self.dataframe.toPandas()["phenotype"])
+        phenotypeCodePairs = list(zip(phenotypes, omimCodes))
```

Comment:
`omniCodes` and `phenotypes` are only created to generate `phenotypeCodePairs` however it is only used to check the pairs at row in row 231. 

I would do something like this:
```Python
for i, (phenotype, code) in self.dataframe.select("omimCode", "phenotype").toPandas().drop_duplicates().iterrows():
```

Or even better, applying `phenotypeCodePairCheck` as an udf on the spark df.

Reply:
I decided to go for the first option and improve this step in the next iteration.